### PR TITLE
DSTOR fixes

### PR DESCRIPTION
**Pull request by karasevb**
* dstor API rework
* dstor access problems resolved

NOTES:
* Still fixing memory corruption on the pmix server side
* in Parallel fixing memory errors with Valgrind, will PR the fixes additionally.

### CODE REVIEW

**Comment by artpol84** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

**Comment by artpol84** [comment]
@karasevb please rebase against the trunk.

**Comment by artpol84** [comment]
bot:retest

**Comment by artpol84** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

**Comment by artpol84** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

**Comment by artpol84** [comment]
bot:retest

**Comment by karasevb** [comment]
bot:retest

File: src/dstore/pmix_esh.c [hunk 1]
```diff
@@ -147,42 +205,41 @@ int _esh_init(void)
...
-        _lockfd = open(_lockfile_name, O_CREAT | O_RDWR | O_EXCL, 0600);
-        /* if previous launch was crashed, the lockfile might not be deleted and unlocked,
-         * so we delete it and create a new one. */
+    if (!_is_server()){
+        _lockfile_name = malloc(PMIX_PATH_MAX);
+        if (_lockfile_name == NULL){
+            PMIX_ERROR_LOG(PMIX_ERROR);
+            return PMIX_ERROR;
+        }
+        snprintf(_lockfile_name, PMIX_PATH_MAX, "%s/%s_dstore_sm.lock", _nspace_path, _unique_id());
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: change to asprintf

File: src/dstore/pmix_esh.c [hunk 2]
```diff
@@ -484,6 +544,72 @@ int _esh_fetch(const char *nspace, int rank, const char *key, pmix_value_t **kvs
...
+    return pmix_setenv(PMIX_DSTORE_ESH_BASE_PATH, _base_path, true, env);
+}
+
+static int _esh_nspace(const char *nspace)
+{
+    struct stat st = {0};
+    char dstore_dir[PMIX_PATH_MAX];
+    seg_desc_t *seg;
+
+    snprintf(dstore_dir, PMIX_PATH_MAX, "%s/%s", _base_path, nspace);
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: `asprintf(_nspace_path, ....)`

File: src/dstore/pmix_esh.c [hunk 3]
```diff
@@ -130,14 +138,64 @@ static inline const char *_unique_id(void)
...
+                    setjobuid = 1;
+                    continue;
+                }
+                if (0 == strcmp(PMIX_DSTPATH, info[n].key)) {
+                    if(_nspace_path != NULL)
+                        free(_nspace_path);
+                    _nspace_path = strdup((char*)info[n].value.data.ptr);
+                    continue;
+                }
+                if (0 == strcmp(PMIX_SERVER_TMPDIR, info[n].key)) {
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: ignore if `PMIX_DSTPATH` was seen

File: src/dstore/pmix_esh.c [hunk 4]
```diff
@@ -484,6 +544,72 @@ int _esh_fetch(const char *nspace, int rank, const char *key, pmix_value_t **kvs
...
+    char dstore_dir[PMIX_PATH_MAX];
+    seg_desc_t *seg;
+
+    snprintf(dstore_dir, PMIX_PATH_MAX, "%s/%s", _base_path, nspace);
+
+    if (stat(dstore_dir, &st) == -1){
+        mkdir(dstore_dir, 0770);
+    }
+
+    if (setjobuid)
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: need brace here

File: src/dstore/pmix_esh.c [hunk 5]
```diff
@@ -484,6 +544,72 @@ int _esh_fetch(const char *nspace, int rank, const char *key, pmix_value_t **kvs
...
+        if (chown(dstore_dir, (uid_t) jobuid, (gid_t) -1) < 0){
+            PMIX_ERROR_LOG(PMIX_ERROR);
+        }
+
+    _nspace_path = strdup(dstore_dir);
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: remove

File: src/server/pmix_server.c [hunk 6]
```diff
@@ -634,6 +635,11 @@ static void _register_nspace(int sd, short args, void *cbdata)
     }
     /* do not destruct the kv object - no memory leak will result */
 
+#if defined(PMIX_ENABLE_DSTORE) && (PMIX_ENABLE_DSTORE == 1)
+    pmix_dstore_nspace_add(cd->proc.nspace);
```
Comment (artpol84) [inline src/server/pmix_server.c]: Address TODO

File: src/server/pmix_server.c [hunk 7]
```diff
@@ -185,10 +185,12 @@ static pmix_status_t initialize_server_base(pmix_server_module_t *module)
     /* find the temp dir */
     if (NULL != mytmpdir) {
         tdir = mytmpdir;
-    } else if (NULL == (tdir = getenv("TMPDIR"))) {
-        if (NULL == (tdir = getenv("TEMP"))) {
-            if (NULL == (tdir = getenv("TMP"))) {
-                tdir = "/tmp";
+    } else if (NULL == (tdir = getenv(PMIX_SERVER_TMPDIR_ENV))) {
```
Comment (artpol84) [inline src/server/pmix_server.c]: Revert, we use info-keys only now in all pmix releases

File: include/pmix_server.h [hunk 8]
```diff
@@ -81,6 +81,7 @@ BEGIN_C_DECLS
  * by the host server via callback function is owned by the host
  * server, which is free to release it upon return from the callback */
 
+#define PMIX_SERVER_TMPDIR_ENV "PMIX_SERVER_TMPDIR"
```
Comment (artpol84) [inline include/pmix_server.h]: Don't need it!

File: src/dstore/pmix_dstore.c [hunk 9]
```diff
@@ -71,3 +73,19 @@ int pmix_dstore_fetch(const char *nspace, int rank, const char *key, pmix_value_
 
     return pmix_dstore.fetch(nspace, rank, key, kvs);
 }
+
+int pmix_dstore_path(char ***env)
```
Comment (artpol84) [inline src/dstore/pmix_dstore.c]: dstor_patch_env

File: src/dstore/pmix_dstore.c [hunk 10]
```diff
@@ -71,3 +73,19 @@ int pmix_dstore_fetch(const char *nspace, int rank, const char *key, pmix_value_
...
+{
+    if (!pmix_dstore.path) {
+        return PMIX_ERR_NOT_SUPPORTED;
+    }
+    return pmix_dstore.path(env);
```
Comment (artpol84) [inline src/dstore/pmix_dstore.c]: .patch

File: src/dstore/pmix_dstore.h [hunk 11]
```diff
@@ -62,8 +64,29 @@ typedef int (*pmix_dstore_base_module_store_fn_t)(const char *nspace, int rank,
...
+/**
+* get base dstore path.
+*
+* @param nspace   namespace string
+*
+* @param rank     rank.
+*
+* @return PMIX_SUCCESS on success.
+*/
+typedef int (*pmix_dstore_base_module_proc_path_fn_t)(char ***env);
```
Comment (artpol84) [inline src/dstore/pmix_dstore.h]: patch_env

File: src/dstore/pmix_dstore.h [hunk 12]
```diff
@@ -73,7 +96,9 @@ typedef struct {
     pmix_dstore_base_module_init_fn_t        init;
     pmix_dstore_base_module_fini_fn_t        finalize;
     pmix_dstore_base_module_store_fn_t       store;
-    pmix_dstrore_base_module_fetch_fn_t      fetch;
+    pmix_dstore_base_module_fetch_fn_t       fetch;
+    pmix_dstore_base_module_proc_path_fn_t   path;
```
Comment (artpol84) [inline src/dstore/pmix_dstore.h]: patch_env

File: src/dstore/pmix_esh.c [hunk 13]
```diff
@@ -28,17 +29,21 @@
...
+static int _esh_nspace(const char *nspace);
 
 pmix_dstore_base_module_t pmix_dstore_esh_module = {
     "esh",
     _esh_init,
     _esh_finalize,
     _esh_store,
-    _esh_fetch
+    _esh_fetch,
+    _esh_path,
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: patch

File: src/dstore/pmix_esh.c [hunk 14]
```diff
@@ -130,14 +138,64 @@ static inline const char *_unique_id(void)
...
     int rc;
     seg_desc_t *seg;
+    size_t n;
 
     PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
                          "%s:%d:%s", __FILE__, __LINE__, __func__));
 
+    /* find the temp dir */
+    if (_is_server()) {
+        if (NULL == (_base_path = getenv(PMIX_SERVER_TMPDIR_ENV))) {
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: No need to check PMIX_SERVER_TMPDIR_ENV

File: src/dstore/pmix_esh.c [hunk 15]
```diff
@@ -130,14 +138,64 @@ static inline const char *_unique_id(void)
...
+        /* scan incoming info for directives */
+        if (NULL != info) {
+            for (n=0; n < ninfo; n++) {
+                if (0 == strcmp(PMIX_USERID, info[n].key)) {
+                    jobuid = info[n].value.data.uint32;
+                    setjobuid = 1;
+                    continue;
+                }
+                if (0 == strcmp(PMIX_DSTPATH, info[n].key)) {
+                    if(_nspace_path != NULL)
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: we set _base_path here as well.
This setting has the priority, we shouldn't ignore PMIX_SERVER_TMPDIR if we already saw PMIX_DSTPATH.

File: src/dstore/pmix_esh.c [hunk 16]
```diff
@@ -147,42 +205,41 @@ int _esh_init(void)
...
 
-    if (_is_server()) {
-        seg = _create_new_segment(INITIAL_SEGMENT, NULL, 0);
-    } else {
-        seg = _attach_new_segment(INITIAL_SEGMENT, NULL, 0);
-    }
-    if (NULL != seg) {
-        _global_sm_seg_first = seg;
-        _global_sm_seg_last = _global_sm_seg_first;
+    if (_is_server()){
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: Can we push it above and avoid 
```C
if (!_is_server()){
```

File: src/dstore/pmix_esh.c [hunk 17]
```diff
@@ -193,15 +250,18 @@ int _esh_finalize(void)
...
     _delete_sm_desc(_global_sm_seg_first);
     PMIX_LIST_DESTRUCT(&_namespace_info_list);
-    if (NULL != _tmp_dir) {
-        free(_tmp_dir);
-    }
+
     close(_lockfd);
     if (_is_server()) {
         unlink(_lockfile_name);
+        rmdir(_nspace_path);
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: check that _nspace_path is != NULL

File: src/dstore/pmix_esh.c [hunk 18]
```diff
@@ -193,15 +250,18 @@ int _esh_finalize(void)
...
     }
+
     free(_lockfile_name);
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]:
```C
if( _lock_file_name != NULL )
```

File: src/dstore/pmix_esh.c [hunk 19]
```diff
@@ -484,6 +544,72 @@ int _esh_fetch(const char *nspace, int rank, const char *key, pmix_value_t **kvs
     return rc;
 }
 
+static int _esh_path(char ***env)
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: patch

File: src/dstore/pmix_esh.c [hunk 20]
```diff
@@ -484,6 +544,72 @@ int _esh_fetch(const char *nspace, int rank, const char *key, pmix_value_t **kvs
...
+        return PMIX_ERROR;
+    }
+
+    return pmix_setenv(PMIX_DSTORE_ESH_BASE_PATH, _base_path, true, env);
+}
+
+static int _esh_nspace(const char *nspace)
+{
+    struct stat st = {0};
+    char dstore_dir[PMIX_PATH_MAX];
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: avoid this additional array

File: src/dstore/pmix_esh.c [hunk 21]
```diff
@@ -484,6 +544,72 @@ int _esh_fetch(const char *nspace, int rank, const char *key, pmix_value_t **kvs
...
+
+    if (setjobuid)
+        if (chown(dstore_dir, (uid_t) jobuid, (gid_t) -1) < 0){
+            PMIX_ERROR_LOG(PMIX_ERROR);
+        }
+
+    _nspace_path = strdup(dstore_dir);
+
+    _lockfile_name = malloc(PMIX_PATH_MAX);
+    snprintf(_lockfile_name, PMIX_PATH_MAX, "%s/%s_dstore_sm.lock", _nspace_path, _unique_id());
```
Comment (artpol84) [inline src/dstore/pmix_esh.c]: asprintf and avoid malloc

**Comment by artpol84** [comment]
bot:retest

**Comment by artpol84** [comment]
:+1:

**Comment by rhc54** [comment]
Can you please setup a PR against the 2.0 branch in pmix-release to bring the DSTOR code up-to-date for release? Thx!

**Comment by artpol84** [comment]
Aren't it already there
https://github.com/pmix/releases/commits/v2.x

**Comment by artpol84** [comment]
I don't see 2.0 branch, I see 2.x. Did you meant that?

**Comment by rhc54** [comment]
Yes, that is what I meant - my bad, you are indeed correct and it has already gone over! I will spin a release candidate when I get home so we can prep a PR for OMPI 2.1.

**Pull request merged**